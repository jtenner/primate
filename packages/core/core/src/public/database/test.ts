export { default } from "#database/test";
