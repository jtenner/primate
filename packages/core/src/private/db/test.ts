import type Database from "#db/Database";
import Store from "#db/Store";
import test from "@rcompat/test";
import type Dict from "@rcompat/type/Dict";
import number from "pema/number";
import optional from "pema/optional";
import primary from "pema/primary";
import string from "pema/string";
import uint from "pema/uint";
import u8 from "pema/u8";

function pick<
  D extends Dict,
  P extends string[],
>(record: D, ...projection: P) {
  return Object.fromEntries(Object.entries(record)
    .filter(([key]) => projection.includes(key)));
};

const users = {
  donald: { name: "Donald", age: 30, lastname: "Duck" },
  ryan: { name: "Ryan", age: 40, lastname: "Wilson" },
  ben: { name: "Ben", age: 60, lastname: "Miller" },
  jeremy: { name: "Just Jeremy", age: 20 },
  paul: { name: "Paul", age: 40, lastname: "Miller" },
} as const;
type User = keyof typeof users;


export default <D extends Database>(database: D) => {
  const _Post = new Store({
    id: primary,
    title: string,
    user_id: uint,
  }, { name: "post", db: database });

  const User = new Store({
    id: primary,
    name: string.default("Donald"),
    lastname: optional(string),
    age: u8,
  }, { name: "user", db: database });

  const bootstrap = async (tester: () => Promise<void>) => {
    await User.schema.create();
    for (const user of Object.values(users)) {
      await User.insert(user);
    }
    await tester();
    await User.schema.delete();
  };

  test.case("insert", async assert => {
    await User.schema.create();

    const donald = await User.insert({ name: "Donald", age: 30 });
    assert(await User.exists(donald.id!)).true();

    const ryan = await User.insert({ name: "Ryan", age: 40 });
    assert(await User.exists(donald.id!)).true();
    assert(await User.exists(ryan.id!)).true();

    await User.schema.delete();
  });

  test.case("find - basic query", async assert => {
    await bootstrap(async () => {
      const result = await User.find({ name: "Ryan" });
      assert(result.length).equals(1);
      assert(result[0]).equals({ id: result[0].id, ...users.ryan });
    });
  });

  test.case("find - sorting by multiple fields", async assert => {
    await bootstrap(async () => {
      // Sorting by multiple fields: age descending, then Lastname ascending
      const sorted = await User.find({}, {
        sort: { age: "desc", lastname: "asc" },
        select: { name: true, age: true },
      });

      assert(sorted.length).equals(5);
      ["ben", "paul", "ryan", "donald", "jeremy"].forEach((user, i) => {
        assert(sorted[i]).equals(pick(users[user as User], "name", "age"));
      });
    });
  });

  test.case("find - sorting ascending and descending", async assert => {
    await bootstrap(async () => {
      const ascending = await User.find({}, {
        sort: { age: "asc" },
        select: { name: true, age: true },
      });

      const ascended = ["jeremy", "donald", "ryan", "paul", "ben"];
      ascended.forEach((user, i) => {
        assert(ascending[i]).equals(pick(users[user as User], "name", "age"));
      });

      const descending = await User.find({}, {
        sort: { age: "desc" },
        select: { name: true, age: true },
      });

      // NB: Since Ryan and Paul have the same age, they will be returned in
      // order of insertion; can't use `ascended.toReversed()`
      const descended = ["ben", "ryan", "paul", "donald", "jeremy"];
      descended.forEach((user, i) => {
        assert(descending[i]).equals(pick(users[user as User], "name", "age"));
      });
    });
  });

  test.case("find - limiting", async assert => {
    await bootstrap(async () => {
      const ascending = await User.find({}, {
        sort: { age: "asc" },
        select: { name: true, age: true },
        limit: 3,
      });

      assert(ascending.length).equals(3);
      const ascended = ["jeremy", "donald", "ryan"];
      ascended.forEach((user, i) => {
        assert(ascending[i]).equals(pick(users[user as User], "name", "age"));
      });
    });
  });

  test.case("update - single record", async assert => {
    await bootstrap(async () => {
      const [donald] = (await User.find({ name: "Donald" }));

      const updated = await User.update(donald.id!, { age: 35 });
      assert(updated.age).equals(35);
      assert(updated).equals({ ...users.donald, id: donald.id!, age: 35 });
    });
  });

  test.case("update - multiple records", async assert => {
    await bootstrap(async () => {
      const updated = await User.update({ age: 40 }, { age: 45 });

      assert(updated.length).equals(2);
      assert(updated[0].age).equals(45);
      assert(updated[1].age).equals(45);
    });
  });

  test.case("delete - single record", async assert => {
    await bootstrap(async () => {
      const [donald] = await User.find({ name: "Donald" });
      await User.delete(donald.id!);

      const deleted = await User.find({ name: "Donald" });
      assert(deleted.length).equals(0);
    });
  });

  test.case("delete - multiple records", async assert => {
    await bootstrap(async () => {
      const n = await User.delete({ age: 40 });
      assert(n).equals(2);

      const remaining = await User.find({});
      assert(remaining.length).equals(3);
      assert(remaining[0].name).equals("Donald");
    });
  });

  test.case("count", async assert => {
    await bootstrap(async () => {
      assert(await User.count()).equals(5);
      assert(await User.count({ name: "Ryan" })).equals(1);
      assert(await User.count({ age: 40 })).equals(2);
      assert(await User.count({ age: 30 })).equals(1);
      assert(await User.count({ age: 35 })).equals(0);
    });
  });

  test.case("exists", async assert => {
    await bootstrap(async () => {
      const [donald] = await User.find({ name: "Donald" });

      assert(await User.exists(donald.id!)).true();
      assert(await User.exists("1234")).false();
    });
  });
};
