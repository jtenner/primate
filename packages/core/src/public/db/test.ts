export { default } from "#db/test";
