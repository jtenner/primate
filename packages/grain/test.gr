module Test
from "runtime/gc" include GC as Runtime

provide let get = () => Some(1)

provide let _decRef = Runtime.decRef
provide let _incRef = Runtime.incRef
