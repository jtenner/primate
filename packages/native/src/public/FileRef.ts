export { default } from "@rcompat/fs/FileRef";
