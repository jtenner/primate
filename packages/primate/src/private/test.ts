import verbs from "@primate/core/verbs";
import type Dict from "@rcompat/type/Dict";

export type Body = string | Dict<string> | Dict<string>[];

export type MockedResponse = {
  status: {
    equals(status: number): void;
  };
  body: {
    equals(body: Body): void;
    includes(body: Body): void;
  };
  headers: {
    includes(headers: Dict<string>): void;
    get(header: string): {
      equals(value: string): void;
      includes(value: string): void;
    };
  };
};

type Verb = typeof verbs[number];

export type Tester = (response: MockedResponse) => void;

export type Route = string;

type Test = {
  verb: Verb;
  route: Route;
  tester: Tester;
};

export const tests: Test[] = [];

export default {
  ...Object.fromEntries(verbs.map(verb =>
    [verb, (route: Route, tester: Tester) => {
      tests.push({ verb, route, tester });
    }],
  )),
} as { [K in Verb]: (path: Route, tester: Tester) => void };
