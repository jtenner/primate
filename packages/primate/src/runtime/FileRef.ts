export { default } from "#runtime/FileRef";
