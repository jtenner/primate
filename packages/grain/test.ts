import FileRef from "@rcompat/fs/FileRef";
import * as wasi from "node:wasi";


const wasiSnapshotPreview1 = new wasi.WASI({
  version: "preview1",
  env: process.env,
  args: process.argv,
});
const wasmPath = "./test.wasm";

const wasm = await WebAssembly.instantiate(
  typeof Bun === "object"
    ? await Bun.file(wasmPath).arrayBuffer()
    : await FileRef.arrayBuffer(wasmPath),
    {
      "wasi_snapshot_preview1": wasiSnapshotPreview1.wasiImport
    }
);

wasiSnapshotPreview1.start(wasm.instance);

for (const [name, _] of Object.entries(wasm.instance.exports)) {
  console.log(name);
}

const result = wasm.instance.exports.get();
const ref = Buffer.from(wasm.instance.exports.memory.buffer, result);
console.log(ref);
