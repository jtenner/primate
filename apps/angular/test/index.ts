import test from "primate/test";
import FileRef from "@rcompat/fs/FileRef";

const expected = await FileRef.text("./index.expected.html");

test.get("/", response => {
  response.body.equals(expected);
});
