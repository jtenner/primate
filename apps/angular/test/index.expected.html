<!doctype html>
<html>
  <head>
    <title>Primate app</title>
    <meta charset="utf-8" />
    <script type="importmap" integrity="sha384-fnSaa3sfY0+f2GD7sEqKe42ysMFFf+cv+q9YsV4jS8EkMttAurz2rPJWMOsdDdfu">{
  "imports": {
    "app": "/app-IFOKCWMT.js"
  }
}</script>
<script type="js" integrity="sha384-l7f5G6DbmthBXrEGm/ST1e6mc7NklBtRd6qiElY0tDHf85yGvi5oB+naGU/1WtMA" src="/app-IFOKCWMT.js"></script>
<script type="js" integrity="sha384-vCM7+AM2lAJzC976iLHKkTU4Sc374LzoykUBHrk6XyJ51SLoIWvMBM7RuVukHHSj" src="/post-index.component.ts.js"></script>
<script type="module" integrity="sha384-nr4CQVLzynYcv7th7OkJBh9OQsSVP7kt9iL5xsgFnqKqjKAoln3THtFKFJfH+8UN">
  import { bootstrapApplication, provideClientHydration } from "app";
  import * as components from "app";

  const config = { providers: [provideClientHydration()] };

  const rendered = components.root_angular(components.angular_175246930aab170eee60n,
    {"posts":[{"id":1,"title":"First post"}]});

  bootstrapApplication(rendered, config)
    .catch(error => console.error(error));</script>

  </head>
  <body><!--nghm--><app-root ng-version="19.2.14" ngh="0" ng-server-context="analog"><post-index ngh="0"><h1>bar</h1><h3>1 says:</h3></post-index></app-root><script id="ng-state" type="application/json">{"__nghData__":[{}]}</script></body>
</html>
